test/e2e/version.go

```diff
@@ -0,0 +1,249 @@
+			versionManager.Update(template, override, clusterNames, versionMap)
+			expectedStatus.ClusterVersions = version.VersionMapToClusterVersions(versionMap)
+			waitForPropVer(tl, client, qualifiedName, expectedStatus)
+		})
+
+		It("should recreate after out-of-band deletion", func() {
+			versionManager.Update(template, override, clusterNames, versionMap)
+			waitForPropVer(tl, client, qualifiedName, expectedStatus)
+
+			// Delete the written version out-of-band
+			err := client.CoreV1alpha1().PropagatedVersions(namespace).Delete(versionName, nil)
+			if err != nil {
+				tl.Fatalf("Error deleting propagated version: %v", err)
+			}
+
+			// Ensure a modified version is written successfully
+			clusterNames, versionMap = removeOneCluster(clusterNames, versionMap)
+			versionManager.Update(template, override, clusterNames, versionMap)
+			expectedStatus.ClusterVersions = version.VersionMapToClusterVersions(versionMap)
+			waitForPropVer(tl, client, qualifiedName, expectedStatus)
+		})
+
+		It("should delete a version from the API", func() {
+			versionManager.Update(template, override, clusterNames, versionMap)
+			waitForPropVer(tl, client, qualifiedName, expectedStatus)
+
+			// Deletion is performed synchronously, so it's not necessary to poll
+			err := versionManager.Delete(util.QualifiedName{Namespace: namespace, Name: template.GetName()})
+			if err != nil {
+				tl.Fatalf("Error deleting propagated version: %v", err)
+			}
+			_, err = client.CoreV1alpha1().PropagatedVersions(namespace).Get(versionName, metav1.GetOptions{})
+			if err == nil {
+				tl.Fatalf("Version manager did not delete propagated version.")
+			} else if err != nil && !errors.IsNotFound(err) {
+				tl.Fatalf("Error retrieving deleted version: %v", err)
+			}
+		})
+	})
+})
+
+func waitForPropVer(tl common.TestLogger, client fedclientset.Interface, qualifiedName util.QualifiedName, expectedStatus fedv1a1.PropagatedVersionStatus) {
+	err := wait.PollImmediate(framework.PollInterval, framework.TestContext.SingleCallTimeout, func() (bool, error) {
+		propVer, err := client.CoreV1alpha1().PropagatedVersions(qualifiedName.Namespace).Get(qualifiedName.Name, metav1.GetOptions{})
+		if errors.IsNotFound(err) {
+			return false, nil
+		}
+		if err != nil {
+			tl.Fatalf("Error retrieving status of propagated version %q: %v", qualifiedName, err)
+		}
+		return util.PropagatedVersionStatusEquivalent(&expectedStatus, &propVer.Status), nil
+	})
+	if err != nil {
+		tl.Fatalf("Timed out waiting for propagated version: %v", err)
+	}
+}
+
+func removeOneCluster(clusterNames []string, versionMap map[string]string) ([]string, map[string]string) {
+	targetName := clusterNames[0]
+	clusterNames[0] = clusterNames[len(clusterNames)-1]
+	clusterNames = clusterNames[:len(clusterNames)-1]
+	delete(versionMap, targetName)
+	return clusterNames, versionMap
+}
+
+const letters = "abcdefghinopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
+
+func randomString(size int) string {
```

Comment:
Is this function necessary? Can we just use the same `uuid.New()` mechanism?

Reply:
I was having trouble using a uuid as a prefix (even though uuid seems fine for a name on its own).  I've experimented with prefixing the uuid with an alpha character (e.g. `v`) and that seems to fix it.